### PR TITLE
REST API: implements getLink by ID service

### PR DESCRIPTION
**Pull request by ArthurHoaro**
See http://shaarli.github.io/api-documentation/#links-link-get

### CODE REVIEW

**Comment by ArthurHoaro** [comment]
Rebased

File: tests/api/controllers/GetLinkIdTest.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+namespace Shaarli\Api\Controllers;
+
+
+use Slim\Container;
+use Slim\Http\Environment;
+use Slim\Http\Request;
+use Slim\Http\Response;
+
+/**
+ * Class LinksTest
```
Comment (virtualtam) [inline tests/api/controllers/GetLinkIdTest.php]: copy-paste spotted :)

File: tests/api/controllers/GetLinkIdTest.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    protected $container;
+
+    /**
+     * @var Links controller instance.
+     */
+    protected $controller;
+
+    /**
+     * Number of JSON field per link.
```
Comment (virtualtam) [inline tests/api/controllers/GetLinkIdTest.php]: *fields

File: tests/api/controllers/GetLinkIdTest.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    const NB_FIELDS_LINK = 9;
+
+    /**
+     * Before every test, instantiate a new Api with its config, plugins and links.
```
Comment (virtualtam) [inline tests/api/controllers/GetLinkIdTest.php]: s/every/each/

File: tests/api/controllers/GetLinkIdTest.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    public function setUp()
+    {
+        $this->conf = new \ConfigManager('tests/utils/config/configJson.json.php');
```
Comment (virtualtam) [inline tests/api/controllers/GetLinkIdTest.php]: `configJson.json.php` => `config.json.php`?

File: tests/api/controllers/GetLinkIdTest.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+        $this->container = new Container();
+        $this->container['conf'] = $this->conf;
+        $this->container['db'] = new \LinkDB(self::$testDatastore, true, false);
+
+        $this->controller = new Links($this->container);
+    }
+
+    /**
+     * After every test, remove the test datastore.
```
Comment (virtualtam) [inline tests/api/controllers/GetLinkIdTest.php]: s/every/each/

File: tests/api/controllers/GetLinksTest.php [hunk 6]
```diff
@@ -16,7 +16,7 @@
  *
  * @package Shaarli\Api\Controllers
  */
-class LinksTest extends \PHPUnit_Framework_TestCase
+class GetLinksTest extends \PHPUnit_Framework_TestCase
```
Comment (virtualtam) [inline tests/api/controllers/GetLinksTest.php]: +docstring update: class name & purpose

File: application/api/controllers/Links.php [hunk 7]
```diff
@@ -83,4 +83,25 @@ public function getLinks($request, $response)
...
+     * @param Response $response Slim response.
+     * @param array    $args     Path parameters. including the ID.
+     *
+     * @return Response containing the link array.
+     *
+     * @throws ApiLinkNotFoundException generating a 404 error.
+     */
+    public function getLink($request, $response, $args)
+    {
+        if (! isset($args['id']) || ! isset($this->linkDb[$args['id']])) {
```
Comment (virtualtam) [inline application/api/controllers/Links.php]: the router already checks the route formatting, so at this point:
- the ID must be non-empty
- it must be an integer

File: tests/api/controllers/GetLinkIdTest.php (hunk 4 above)
Comment (ArthurHoaro) [inline tests/api/controllers/GetLinkIdTest.php]: No, that's its name (created during configJson/configPhp upgrade tests). However, the extension shouldn't be included.

**Comment by ArthurHoaro** [comment]
Thank you!

**Pull request merged**